drivers: adc: nrfx: enable negative values for single-ended ADC readings

jaz1-nordic opened the pull request:
Based on #91368, only the last commit is in scope of this PR.

Add custom nordic-specific ADC API to enable to read negative values from a single-ended ADC channels.

nika-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -725,6 +727,22 @@ static int adc_nrfx_read(const struct device *dev,
 {
 	int error;
 
+	m_data.correct_single_ended = true;
```
should probably be done within `context_lock`, otherwise there is a risk of race condition. Same below

nika-nordic commented on include/zephyr/drivers/adc/adc_nrfx.h:
```diff
@@ -0,0 +1,48 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_ADC_ADC_NRFX_H_
```
I suggest to use `adc_nrf_*` namespace (for file and function names) to not imply any specific implementation underneath vendor-specific API.

jaz1-nordic committed:
drivers: adc: nrfx: enable negative values for single-ended ADC readings

Add custom nordic-specific ADC API to enable to read negative values
from a single-ended ADC channels.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic committed:
drivers: adc: nrfx_saadc: Add support for SAADC internal sampling timer

The SAMPLERATE register can be used as a local timer instead
of triggering individual SAMPLE tasks. When SAMPLERATE.MODE is set
to Timers, it is sufficient to trigger SAMPLE task only once in order
to start the SAADC and triggering the STOP task will stop sampling.
The SAMPLERATE.CC field controls the sample rate.

The SAMPLERATE timer should not be combined with SCAN mode and
only one channel should be enabled when using the internal timer.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic committed:
drivers: adc: nrfx_saadc: Add support for DMM

Add support for DMM which manages cache and dedicated memory spaces.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

bjarki-andreasen reviewed (changes requested):
The ADC driver API already supports ADC readings which can return signed values, these are differential readings. In Nordic's datasheet, we have a mode called "single ended", but like the RTC (which is actually a lower power counter, not a real time clock as zephyr defines it), its just a name. "'single ended" is a differential reading, with the negative channel tied to GND. This is not compatible with zephyrs definition of a single ended reading.

To support Nordic's "single ended" mode, the user must configure a differential reading, with the negative input tied to ground:
```
	channel@7 {
		reg = <7>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,input-positive = <NRF_SAADC_AIN6>; /* P1.13 */
		zephyr,input-negative = <NRF_SAADC_GND>; /* <<< new */
		zephyr,resolution = <12>;
	};
```
which the saadc driver can then use to configure the reading as Nordic SAADC "single ended", and return negative values as expected :)

jaz1-nordic commented:
Handled here:
https://github.com/zephyrproject-rtos/zephyr/pull/94069
